storage_controller/src/service.rs

```diff
@@ -3116,6 +3081,81 @@ impl Service {
         results
     }
 
+    /// Helper for safely working with the shards in a tenant remotely on pageservers, for example
+    /// when creating and deleting timelines:
+    /// - Makes sure shards are attached somewhere if they weren't already
+    /// - Looks up the shards and the nodes where they were most recently attached
+    /// - Guarantees that after the inner function returns, the shards' generations haven't moved on: this
+    ///   ensures that the remote operation acted on the most recent generation, and is therefore durable.
+    async fn tenant_remote_mutation<R, O, F>(
+        &self,
+        tenant_id: TenantId,
+        op: O,
+    ) -> Result<R, ApiError>
+    where
+        O: FnOnce(Vec<(TenantShardId, Node)>) -> F,
+        F: std::future::Future<Output = R>,
+    {
+        let target_gens = {
+            let mut targets = Vec::new();
+
+            // Load the currently attached pageservers for the latest generation of each shard.  This can
+            // run concurrently with reconciliations, and it is not guaranteed that the node we find here
+            // will still be the latest when we're done: we will check generations again at the end of
+            // this function to handle that.
+            let generations = self.persistence.peek_generations(tenant_id).await?;
+            let generations = if generations.iter().any(|i| i.1.is_none()) {
+                // One or more shards is not attached to anything: maybe this is a new tenant?  Wait for
+                // it to reconcile.
+                self.ensure_attached_wait(tenant_id).await?;
+                self.persistence.peek_generations(tenant_id).await?
```

Comment:
Nothing prevents a new reconiliation or shard split to start between these two lines.

I think what we _want_ is to invoke `op` with exactly the reconciled view at the time of `ensure_attached_wait`.

Can we return that view from `ensure_attached_wait`?

That would also allow us to have always `Some(generation)`.

Reply:
>Nothing prevents a new reconiliation or shard split to start between these two lines.

It's not obvious in this context, but callers always hold a lock on the tenant that prevents splits from happening.

Concurrent reconciliations are permitted: if one of them happens with timing that violates our generation-number-stays-constant requirements, we'll 503 and the caller will retry.

> I think what we want is to invoke op with exactly the reconciled view at the time of ensure_attached_wait.

It's intentional to go to the database rather than trust the output of the last reconcile: if there is some other reconcile running in the background, it might have already attached a new pageserver.

Reply:
I'm going to adjust this piece of code in https://github.com/neondatabase/neon/issues/8819, I'm not sure we even need the ensure_attached_wait part at all any more.